library/core/src/cell.rs

```diff
@@ -1766,15 +1766,19 @@ impl<T: ?Sized + fmt::Display> fmt::Display for RefMut<'_, T> {
 ///
 /// The precise Rust aliasing rules are somewhat in flux, but the main points are not contentious:
 ///
-/// - If you create a safe reference with lifetime `'a` (either a `&T` or `&mut T`
-/// reference) that is accessible by safe code (for example, because you returned it),
-/// then you must not access the data in any way that contradicts that reference for the
-/// remainder of `'a`. For example, this means that if you take the `*mut T` from an
-/// `UnsafeCell<T>` and cast it to an `&T`, then the data in `T` must remain immutable
-/// (modulo any `UnsafeCell` data found within `T`, of course) until that reference's
-/// lifetime expires. Similarly, if you create a `&mut T` reference that is released to
-/// safe code, then you must not access the data within the `UnsafeCell` until that
-/// reference expires.
+/// - If you create a safe reference with lifetime `'a` (either a `&T` or `&mut T` reference), then
+/// you must not access the data in any way that contradicts that reference for the remainder of
+/// `'a`. For example, this means that if you take the `*mut T` from an `UnsafeCell<T>` and cast it
+/// to an `&T`, then the data in `T` must remain immutable (modulo any `UnsafeCell` data found
+/// within `T`, of course) until that reference's lifetime expires. Similarly, if you create a `&mut
+/// T` reference that is released to safe code, then you must not access the data within the
+/// `UnsafeCell` until that reference expires.
+///
+/// - For both `&T` without `UnsafeCell<T>` and `&mut T`, you must also not deallocate the data
+/// until the reference expires. As a special exception, `&UnsafeCell<T>` references may be
```

Comment:
It's a bit hard for me to tell from reading this whether it refers to `&UnsafeCell<T>` *precisely* or "any type that contains an unsafe cell" etc. From private discussion, I have the impression that it is meant to mean "given an `&T`, any part of `T` that is owned by an `UnsafeCell` may be deallocated or mutated, but parts of `T` that are not owned by an `UnsafeCell` cannot be". Since you can't deallocate "part" of a memory space, that implies that `&UnsafeCell<T>` can be deallocated by the tuple referenced by `&(u32, UnsafeCell<i32>)` cannot be, right?

This generalized form makes sense to me and is clearer than saying "also we special case this particular type", which is how I read this paragraph.

Is that what is intended?

Reply:
> "given an &T, any part of T that is owned by an UnsafeCell may be deallocated or mutated, but parts of T that are not owned by an UnsafeCell cannot be"

Yes, that is the intended interpretation. This is basically the smallest change that I think is reasonable and what makes `Arc` correct as-written. I personally would be fine with going further and saying the `UnsafeCell` infects the entire reference (for both mutation and deallocation), but that is a much more contentious change (https://github.com/rust-lang/unsafe-code-guidelines/issues/236).

> Since you can't deallocate "part" of a memory space, that implies that &UnsafeCell<T> can be deallocated by the tuple referenced by &(u32, UnsafeCell<i32>) cannot be, right?

Correct. Worse, a `&(UnsafeCell<i16>, UnsafeCell<i32>)` can also not be deallocated since the padding is outside the `UnsafeCell`. I admit this is quite the footgun.

I'm happy to clarify the wording. I am not sure what is the best way to say this.

Reply:
@nikomatsakis that said, we could also for now just say that the reference literally has to be `&UnsafeCell<T>` (or some newtype around `UnsafeCell`). That would still be enough for the cases I know. It is more restrictive than what Miri implements but this is not the only case where the docs are hedging our bets more carefully than Miri.

I'm fine with either approach; not sure what y'all would prefer.

Reply:
Personally I prefer the less-narrow version, it seems more like a coherent rule and less like a hack. =)